src/autoschedulers/anderson2021/LoopNest.cpp

```diff
@@ -0,0 +1,4001 @@
+    }
+
+    *working_set += working_set_here;
+}
+
+void LoopNest::recompute_inlined_features(const StageMap<Sites> &sites, StageMap<ScheduleFeatures> *features) const {
+    for (const auto &c : children) {
+        c->recompute_inlined_features(sites, features);
+    }
+
+    for (auto it = inlined.begin(); it != inlined.end(); it++) {
+        const auto *f = it.key();
+        internal_assert(f);
+
+        const auto &block = sites.get(stage).task;
+
+        internal_assert(sites.contains(block->stage));
+        uint64_t hash_of_producers = sites.get(block->stage).hash_of_producers_stored_at_root;
+
+        internal_assert(block->feature_intermediates.count(hash_of_producers) > 0);
+        auto &intermediate_map = block->feature_intermediates[hash_of_producers].get(&(f->stages[0]));
+        auto &intermediate = intermediate_map.get(stage);
+
+        auto &inlined_feat = features->get(&(f->stages[0]));
+        inlined_feat.inlined_calls += intermediate.inlined_calls;
+        inlined_feat.num_scalars += intermediate.num_scalars;
+        if (inlined_feat.innermost_pure_loop_extent > 0) {
+            inlined_feat.innermost_pure_loop_extent =
+                std::min(inlined_feat.innermost_pure_loop_extent,
+                         intermediate.innermost_pure_loop_extent);
+        } else {
+            inlined_feat.innermost_pure_loop_extent = intermediate.innermost_pure_loop_extent;
+        }
+        inlined_feat.outer_parallelism = intermediate.outer_parallelism;
+        inlined_feat.num_blocks = intermediate.outer_parallelism;
+        inlined_feat.num_warps_per_block += intermediate.num_warps_per_block;
+
+        inlined_feat.num_threads_per_block += intermediate.num_threads_per_block;
+        inlined_feat.points_computed_per_thread += intermediate.points_computed_per_thread;
+    }
+}
+
+std::pair<int64_t, bool> LoopNest::compute_alloc_size_of_node_here(const FunctionDAG::Node *f) const {
+    const auto &bounds = get_bounds(f);
+
+    auto bytes = f->bytes_per_point;
+    bool is_constant = true;
+    for (int i = 0; i < f->dimensions; i++) {
+        const auto &p = bounds->region_computed(i);
+        bytes *= p.extent();
+        is_constant = is_constant && p.constant_extent();
+    }
+
+    return {bytes, is_constant};
+}
+
+// Do a recursive walk over the loop nest computing features to feed the cost model.
+void LoopNest::compute_features(const FunctionDAG &dag,
+                                const Anderson2021Params &params,
+                                const Target &target,
+                                const StageMap<Sites> &sites,
+                                int64_t instances,
+                                int64_t parallelism,
+                                const LoopNest *parent,
+                                const LoopNest *grandparent,
+                                const LoopNest &root,
+                                int64_t *working_set,
```

Comment:
It's unusual to have out-params stuck in the middle of the argument list (rather than at the end).